File: atividade-13/atividade.py
class AFD:
    def __init__(self):
        self.state = 'q0'  

    def transition(self, char):
        if self.state == 'q0':
            if char == '1':
                self.state = 'q1'
            elif char == '0':
                self.state = 'q-1' 
        elif self.state == 'q1':
            if char == '1':
                self.state = 'q2' 
            elif char == '0':
                self.state = 'q0'
        elif self.state == 'q-1':
            if char == '1':
                self.state = 'q0' 
            elif char == '0':
                self.state = 'q-2' 
        elif self.state == 'q2':
            if char == '1':
                self.state = 'q3'
            elif char == '0':
                self.state = 'q1' 
        elif self.state == 'q-2':
            if char == '1':
                self.state = 'q-1' 
            elif char == '0':
                self.state = 'q-3' 

    def is_accepted(self, string):
        self.state = 'q0' 
        for char in string:
            self.transition(char)

        return self.state in {'q1', 'q2', 'q3'}

if __name__ == '__main__':
    afd = AFD()
    print(afd.is_accepted("110"))   
    print(afd.is_accepted("101"))   
    print(afd.is_accepted("11100"))
    print(afd.is_accepted("100"))   
    print(afd.is_accepted("111"))   

File: atividade-13/atividade.go
package main

import "fmt"

type AFD struct {
	state string
}

func NewAFD() *AFD {
	return &AFD{state: "q0"}
}

func (afd *AFD) transition(input rune) {
	switch afd.state {
	case "q0":
		if input == '1' {
			afd.state = "q1"
		} else if input == '0' {
			afd.state = "q-1"
		}
	case "q1":
		if input == '1' {
			afd.state = "q2"
		} else if input == '0' {
			afd.state = "q0"
		}
	case "q-1":
		if input == '1' {
			afd.state = "q0"
		} else if input == '0' {
			afd.state = "q-2"
		}
	case "q2":
		if input == '1' {
			afd.state = "q3"
		} else if input == '0' {
			afd.state = "q1"
		}
	case "q-2":
		if input == '1' {
			afd.state = "q-1"
		} else if input == '0' {
			afd.state = "q-3"
		}
	}
}

func (afd *AFD) isAccepted(input string) bool {
	afd.state = "q0"
	for _, c := range input {
		afd.transition(c)
	}


	return afd.state == "q1" || afd.state == "q2" || afd.state == "q3"
}

func main() {
	afd := NewAFD()
	fmt.Println(afd.isAccepted("110"))  
	fmt.Println(afd.isAccepted("101"))  
	fmt.Println(afd.isAccepted("11100"))
	fmt.Println(afd.isAccepted("100"))  
	fmt.Println(afd.isAccepted("111"))  
}
